providers/os/resources/firewalld.go

```diff
@@ -0,0 +1,1309 @@
+	}
+
+	return firewalldZoneData{
+		Zone:  zone,
+		Rules: rules,
+	}, nil
+}
+
+func interfacesToStrings(items []firewalldInterfaceXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		val := strings.TrimSpace(item.Name)
+		if val != "" {
+			res = append(res, val)
+		}
+	}
+	return res
+}
+
+func sourcesToStrings(items []firewalldSourceXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		val := strings.TrimSpace(item.Address)
+		if val != "" {
+			res = append(res, val)
+		}
+	}
+	return res
+}
+
+func namesToStrings(items []firewalldNameXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		val := strings.TrimSpace(item.Name)
+		if val != "" {
+			res = append(res, val)
+		}
+	}
+	return res
+}
+
+func portsToStrings(items []firewalldPortXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		port := strings.TrimSpace(item.Port)
+		proto := strings.TrimSpace(item.Protocol)
+		switch {
+		case port == "" && proto == "":
+			continue
+		case port != "" && proto != "":
+			res = append(res, port+"/"+proto)
+		default:
+			res = append(res, port+proto)
+		}
+	}
+	return res
+}
+
+func sourcePortsToStrings(items []firewalldPortXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		port := strings.TrimSpace(item.Port)
+		proto := strings.TrimSpace(item.Protocol)
+		if port == "" && proto == "" {
+			continue
+		}
+
+		entry := []string{}
+		if port != "" {
+			entry = append(entry, "port="+port)
+		}
+		if proto != "" {
+			entry = append(entry, "proto="+proto)
+		}
+		if len(entry) > 0 {
+			res = append(res, strings.Join(entry, ":"))
+		}
+	}
+	return res
+}
+
+func forwardPortsToStrings(items []firewalldForwardPortXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		parts := []string{}
+		if v := strings.TrimSpace(item.Port); v != "" {
+			parts = append(parts, "port="+v)
+		}
+		if v := strings.TrimSpace(item.Protocol); v != "" {
+			parts = append(parts, "proto="+v)
+		}
+		if v := strings.TrimSpace(item.ToPort); v != "" {
+			parts = append(parts, "toport="+v)
+		}
+		if v := strings.TrimSpace(item.ToAddr); v != "" {
+			parts = append(parts, "toaddr="+v)
+		}
+		if len(parts) > 0 {
+			res = append(res, strings.Join(parts, ":"))
+		}
+	}
+	return res
+}
+
+func protocolsToStrings(items []firewalldProtocolXML) []string {
+	if len(items) == 0 {
+		return nil
+	}
+	res := make([]string, 0, len(items))
+	for _, item := range items {
+		val := strings.TrimSpace(item.Value)
+		if val != "" {
+			res = append(res, val)
+		}
+	}
+	return res
+}
+
+func buildRichRuleString(rule firewalldRuleXML) string {
```

Comment:
should also be called `ToTokens` to align with `func (e *firewalldRuleEndpointXML) toTokens() []string {}`